README
NAME
    Math::Base::Convert - very fast base to base conversion

SYNOPSIS
  As a function

      use Math::Base::Convert qw( :all )
      use Math::Base::Convert qw( 

            cnv
            cnvabs
            cnvpre
            basemap

                            # comments
            bin             base 2 0,1
            dna             base 4 lower case dna
            DNA             base 4 upper case DNA
            oct             base 8 octal
            dec             base 10 decimal
            hex             base 16 lower case hex
            HEX             base 16 upper case HEX
            b62             base 62
            b64             base 64 month:C:12 day:V:31
            m64             base 64 0-63 from MIME::Base64
            iru             base 64 P10 protocol - IRCu daemon
            url             base 64 url with no %2B %2F expansion of + - /
            rex             base 64 regular expression varient
            id0             base 64 IDentifier style 0
            id1             base 64 IDentifier style 1
            xnt             base 64 XML Name Tokens (Nmtoken)
            xid             base 64 XML identifiers (Name)
            b85             base 85 RFC 1924 for IPv6 addresses
            ascii           base 96 7 bit printible 0x20 - 0x7F
      );

      my $converted = cnv($number,optionalFROM,optionalTO);
      my $basemap = basmap(base);

  As a method:

      use Math::Base::Convert;
      use Math::Base::Convert qw(:base);

      my $bc = new Math::Base::Convert(optionalFROM,optionalTO);
      my $converted = $bc->cnv($number);
      my $basemap = $bc->basemap(base);

DESCRIPTION
    This module provides fast functions and methods to convert between
    arbitrary number bases from 2 (binary) thru 65535.

    This module is pure Perl, has no external dependencies, and is backward
    compatible with old versions of Perl 5.

PREFERRED USE
    Setting up the conversion parameters, context and error checking consume
    a significant portion of the execution time of a single base conversion.
    These operations are performed each time cnv is called as a function.

    Using method calls eliminates a large portion of this overhead and will
    improve performance for repetitive conversions. See the benchmarks
    sub-directory in this distribution.

BUILT IN NUMBER SETS
    Number set varients courtesy of the authors of Math::Base:Cnv and
    Math::BaseConvert.

    The functions below return a reference to an array

      $arrayref     = function;

      bin => ['0', '1']                               # binary
      dna => ['a','t','c','g']                        # lc dna
      DNA => ['A','T','C','G'],     {default}         # uc DNA
      oct => ['0'..'7']                               # octal
      dec => ['0'..'9']                               # decimal
      hex => ['0'..'9', 'a'..'f']                     # lc hex
      HEX => ['0'..'9', 'A'..'F']   {default}         # uc HEX
      b62 => ['0'..'9', 'a'..'z', 'A'..'Z']           # base 62
      b64 => ['0'..'9', 'A'..'Z', 'a'..'z', '.', '_'] # m:C:12 d:V:31
      m64 => ['A'..'Z', 'a'..'z', '0'..'9', '+', '/'] # MIMI::Base64
      iru => ['A'..'Z', 'a'..'z', '0'..'9', '[', ']'] # P10 - IRCu
      url => ['A'..'Z', 'a'..'z', '0'..'9', '*', '-'] # url no %2B %2F
      rex => ['A'..'Z', 'a'..'z', '0'..'9', '!', '-'] # regex varient
      id0 => ['A'..'Z', 'a'..'z', '0'..'9', '_', '-'] # ID 0
      id1 => ['A'..'Z', 'a'..'z', '0'..'9', '.', '_'] # ID 1
      xnt => ['A'..'Z', 'a'..'z', '0'..'9', '.', '-'] # XML (Nmtoken)
      xid => ['A'..'Z', 'a'..'z', '0'..'9', '_', ':'] # XML (Name)
      b85 => ['0'..'9', 'A'..'Z', 'a'..'z', '!', '#', # RFC 1924
              '$', '%', '&', '(', ')', '*', '+', '-', 
              ';', '<', '=', '>', '?', '@', '^', '_', 
              '', '{', '|', '}', '~']
      An arbitrary base 96 composed of printable 7 bit ascii
      from 0x20 (space) through 0x7F (tilde ~)
      ascii => [
            ' ','!','"','#','$','%','&',"'",'(',')',
            '*','+',',','-','.','/',
            '0','1','2','3','4','5','6','7','8','9',
            ':',';','<','=','>','?','@',
            'A','B','C','D','E','F','G','H','I','J','K','L','M',
            'N','O','P','Q','R','S','T','U','V','W','X','Y','Z',
            '[','\',']','^','_','`',
            'a','b','c','d','e','f','g','h','i','j','k','l','m',
            'n','o','p','q','r','s','t','u','v','w','x','y','z',
            '{','|','}','~']

      NOTE: Clean text with =~ s/\s+/ /; before applying to ascii

USAGE
    * $converted = cnv($number,[from],[to])
        SCALAR context: array context covered later in this document.

        To preserve similarity to other similar base conversion modules, cnv
        returns the converted number string with SIGN if both the input and
        output base strings are in known signed set of bases in this module.

        In the case of binary, octal, hex, all leading base designator
        strings such as '0b','0', '0x' are automatically stripped from the
        input. Base designator strings are NOT applied to the output.

        The context of base FROM and TO is optional and flexible.

        Unconditional conversion from decimal to HEX [upper case]

                $converted = cnv($number);

        Example conversion from octal to default HEX [upper case] with
        different context for the 'octal' designator.

          base as a number
                $converted = cnv($number,8);

          base as a function    (imported)
                $converted = cnv($number,oct);

          base as text
                $converted = convbase($number,'oct');

        Conversion to/from arbitary bases i.e.

          $converted = cnv($number); # dec -> hex (default)
          $converted = cnv($number,oct);        # oct to HEX
          $converted = cnv($number,10,HEX);     # dec to uc HEX
          $converted = cnv($number,10,hex);     # dec to lc hex
          $converted = cnv($number,dec,hex);#    same

                pointer notation
          $converted = cnv($number, oct => dec);

          $converted = cnv($number,10 => 23); # dec to base23
          $converted = cnv($number,23 => 5);  # b23 to base5
          etc...

    * $bc = new Math::Base::Convert([from],[to]);
        This method has the same usage and syntax for FROM and TO as cnv
        above.

        Setup for unconditional conversion from HEX to decimal

                $bc = new Math::Base::Convert();

        Example conversion from octal to decimal

          base number
                $bc = new Math::Base::Convert(8);

          base function (imported)
                $bc = new Math::Base::Convert(oct);

          base text
                $bc = new Math::Base::Convert('oct')

        The number conversion for any of the above:

        NOTE: iterative conversions using a method pointer are ALWAYS faster
        than calling cnv as a function.

                $converted = $bc->cnv($number);

    * $converted = cnvpre($number,[from],[to])
        Same as cnv except that base descriptor PREfixes are applied to
        binary, octal, and hexadecimal output strings.

    * $converted = cnvabs($number,[from],[to])
        Same as cnv except that the ABSolute value of the number string is
        returned without SIGN is returned. i.e. just the raw string.

    * ($sign,$prefix,$string) = cnv($number,[$from,[$to]])
    * ($sign,$prefix,$string) = cnv($number,[$from,[$to]])
    * ($sign,$prefix,$string) = cnv($number,[$from,[$to]])
        ARRAY context:

        All three functions return the same items in array context.

          sign          the sign of the input number string

          prefix        the prefix which would be applied to output

          string        the raw output string

    * $basemap = basemap(base);
    * $basemap = $bc->basemap(base);
        This function / method returns a pointer to a hash that maps the
        keys of a base to its numeric value for base conversion. It accepts
        base in any of the forms described for cnv.

        The return basemap includes upper and lower case varients of the the
        number base in cases such as hex where upper and lower case a..f,
        A..F map to the same numeric value for base conversion.

          i.e. $hex_ptr = {
                0  => 0,
                1  => 1,
                2  => 2,
                3  => 3,
                4  => 4,
                5  => 5,
                6  => 6,
                7  => 7,
                8  => 8,
                9  => 9,
                A  => 10,
                B  => 11,
                C  => 12,
                D  => 13,
                E  => 14,
                F  => 15,
                a  => 10,
                b  => 11,
                c  => 12,
                d  => 13,
                e  => 14,
                f  => 15
          };

BENCHMARKS
    Math::Base::Convert includes 2 development and one real world benchmark
    sequences included in the test suite. Benchmark results for a 500mhz
    system can be found in the 'benchmarks' source directory.

      make test BENCHMARK=1

    Provides comparison data for bi-directional conversion of an ascending
    series of number strings in all base powers. The test sequence contains
    number strings that go from a a single 32 bit register to several.
    Tested bases are: (note: b32, b128, b256 not useful and are for testing
    only)

        base 2    4    8    16   32   64   85   128   256
            bin, dna, oct, hex, b32, b64, b85, b128, b256

    Conversions are performed FROM all bases TO decimal and are repeated in
    the opposing direction FROM decimal TO all bases.

    Benchmark 1 results indicate the Math::Base::Convert typically runs
    significantly faster ( 10x to 100x) than Math::BigInt based
    implementations used in similar modules.

      make test BENCHMARK=2

    Provides comparison data for the frontend and backend converters in
    Math::Base::Convert's CalcPP and Shortcuts packages, and Math::Bigint
    conversions if it is present on the system under test.

      make test BENCHMARK=3

    Checks the relative timing of short and long number string conversions.
    FROM a base number to n*32 bit register and TO a base number from an
    n*32 bit register set.

    i.e. strings that convert to and from 1, 2, 3... etc.. 32 bit registers

DEPENDENCIES
            none

            Math::BigInt is conditionally used in
            the test suite but is not a requirement

EXPORT_OK
    Conditional EXPORT functions

            cnv
            cnvabs
            cnvpre
            basemap
            bin
            oct
            dec
            heX
            HEX
            b62
            b64
            m64
            iru
            url
            rex
            id0
            id1
            xnt
            xid
            b85
            ascii

EXPORT_TAGS
    Conditional EXPORT function groups

            :all    => all of above
            :base   => all except 'cnv,cnvabs,cnvpre'

ACKNOWLEDGEMENTS
    This module was inspired by Math::BaseConvert maintained by Shane Warden
    <chromatic@cpan.org> and forked from Math::BaseCnv, both authored by Pip
    Stuart <Pip@CPAN.Org>

AUTHOR
    Michael Robinton, <miker@cpan.org>

COPYRIGHT
    Copyright 2012-2014, Michael Robinton

    This program is free software; you may redistribute it and/or modify it
    under the same terms as Perl itself.

    This program is distributed in the hope that it will be useful, but
    WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.



Changes
Revision history for Perl extension Math::Base::Convert

0.08  Mon Jan 27 11:59:58 PST 2014
        modify Makefile.PL to bypass missing 'pod2text'

	correct documentation typo

0.07  Tue Jan  8 13:47:58 PST 2013
  revise Shortcuts 0.02 srindx table use to mask after shift

  floating point error detected ONLY in 
	fail   2012-12-30T20:29:47 v5.16.2/127e
	fail   2013-01-03T03:49:32 v5.12.5/a2da
	fail   2013-01-03T05:24:45 v5.14.3/127e
	fail   2013-01-03T05:43:01 v5.17.6/127e
  64 bit perl compiled with -Duselongdouble

  wrap base calculation with int( ... +0.5) in Shortcuts.pm
  $bp = int(log($base/log(2) + 0.5)

  MANY thanks to Andreas Koenig <andreas.koenig.7os6VVqR@franz.ak.mind.de>
  for all the help he provide in tracking this bug down and to 
  David Cantrell <david@cantrell.org.uk> for providing a guest account
  on his system to do the testing.

0.06  Wed Dec 26 00:08:52 PST 2012
  changed Bases.pm array returns from arrays built in the sub
  to arrays built in the package and already blessed scalars

  add ascii lookup table

0.05  Sun Dec 23 22:46:42 PST 2012
  initial release

0.04  Sat Dec 22 21:07:56 PST 2012
  \Q \E metaquoting escape sequence to disable all 
  metacharacters in later versions of Perl

0.03  Tue Dec 18 17:44:58 PST 2012
  archive before TO <-> FROM switch

0.02  Wed Dec  5 17:56:47 PST 2012
  separated math into CalcPP.pm

0.01  Wed Dec  5 13:59:41 PST 2012
  first development package


Makefile.PL
use ExtUtils::MakeMaker;
use Config;

unless ( eval {require './bitmaps'} ) {
  print STDERR "Makefile.PL invoked the 'bitmaps' script\nwhich failed to create the module\nlib/Math/Base/Convert/Bitmaps.pm\n";
  exit 1;
}

my $pkg = 'Math::Base::Convert';
$pkg =~ /[^:]+$/;
my $module = $& .'.pm';

# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.

WriteMakefile(
    'NAME'		=> $pkg,
    'VERSION_FROM'	=> $module,			# finds $VERSION
    'clean'		=> { FILES    => "*~ tmp* lib/Math/Base/Convert/*~" },
    'dist'		=> {COMPRESS=>'gzip', SUFFIX=>'gz'}
);

sub MY::top_targets {
  package MY;
  my $inherited = shift->SUPER::top_targets(@_);
  $inherited =~ s/(pure_all\s+::.+)/$1 README/;
  $inherited;
}

sub MY::post_constants {
  my $post_constants = q|
MY_POD2TEXT = |. $Config{scriptdirexp} .'/pod2text' .q|   
|;
} 

sub MY::postamble {
  package MY;
  my $postamble = q|
README  : |. $module .q|
	@$(MY_POD2TEXT) |. $module .q| > README

|;
}


lib/Math/Base/Convert/CalcPP.pm
#!/usr/bin/perl

package Math::Base::Convert::CalcPP;

use strict;
use vars qw($VERSION);

$VERSION = do { my @r = (q$Revision: 0.02 $ =~ /\d+/g); sprintf "%d."."%02d" x $#r, @r };

# test number < 2^32 is NOT power of 2
#
sub isnotp2 {
  my $ref = ref $_[0];
  shift if ref $_[0] || $_[0] =~ /\D/;	# class?
  $_[0] & $_[0] -1;
}

# add a long n*32 bit number toa number < 65536
# add 'n' to array digits and propagate carry, return carry
#
sub addbaseno {
  my($ap,$n) = @_;
  foreach (@$ap) {
    $_ += $n;
    return 0 unless $_ > 0xffffffff;
    $n = 1;
    $_ -= 4294967296;
  }
  1;	# carry is one on exit, else would have taken return 0 branch
}

# multiply a register of indeterminate length by a number < 65535
#
# ap    pointer to multiplicand array
# multiplier  
#
sub multiply {
  my($ap,$m) = @_;
# $m is always 2..65535
#  $m &= 0xffff;                # max value 65535       already done by VETTING
#
# perl uses doubles for arithmetic, $m << 65536 will fit
  my $carry = 0;
  foreach ( @$ap) {
    $_ *= $m;
    $_ += $carry;
    if ($_ > 0xffffffff) {
      $carry = int($_ / 4294967296);
      $_ %= 4294967296;
    } else {
      $carry = 0;
    }
  }
  push @$ap, $carry if $carry;
}

sub dividebybase {
  my($np,$divisor) = @_;
  my @dividend = @$np;		# 3% improvement
  while ($#dividend) {		# 3% improvement
    last if $dividend[0];
    shift @dividend;
  }    
  my $remainder = 0;
  my @quotient;
  while (@dividend) {
    my $work = ($dividend[0] += ($remainder * 4294967296));     
    push @quotient, int($work / $divisor);
    $remainder = $work % $divisor;
    shift @dividend;
  }
  return (\@quotient,$remainder);
}

# simple versions of conversion, works for N < ~2^49 or 10^16
#
#sub frombase {
#  my($hsh,$base,$str) = @_;
#  my $number = 0;
#  for( $str =~ /./g ) {
#    $number *= $base;
#    $number += $hsh->{$_};
#  }
#  return $number;
#}

#sub tobase {
#sub to_base
#  my($bp,$base,$num) = @_;
#  my $base   = shift;
#  return $bp->[0] if $num == 0;
#  my $str = '';
#  while( $num > 0 ) {
#    $str = $bp->[$num % $base] . $str;
#    $num = int( $num / $base );
#  }
#  return $str;
#}

# convert a number from its base to 32*N bit representation
#
sub useFROMbaseto32wide {
  my $bc = shift;
  my($ary,$hsh,$base,$str) = @{$bc}{qw(from fhsh fbase nstr)};
# check if decimal and interger from within perl's 32bit double representation
# cutoff is 999,999,999,999,999 -- a bit less than 2^50
#
#	convert directly to base 2^32 arrays
#
  my @result = (0);

  if ($base == 10 && length($str) < 16) {
#  unless ($str > 999999999999999) {	# maximum 32 bit double float integer representation
    $result[0] = $str % 4294967296;
    my $quotient = int($str / 4294967296);
    $result[1] = $quotient if $quotient;
    $bc->{b32str} = \@result;
  }
  else {
    for ($str =~ /./g) {
      multiply(\@result,$base);
      push @result, 1 if addbaseno(\@result,$hsh->{$_});	# propagate carry
    }
#    my @rv = reverse @result;
    $bc->{b32str} = \@result;
  }
  $bc;
}

#my %used = map {$_,0}(0..255);

# convert 32*N bit representation to any base < 65536
#

sub use32wideTObase {
  my $bc = shift;
  my($ary,$base,$rquot) = @{$bc}{qw(to tbase b32str)};
  my @quotient = reverse(@$rquot);
  my $quotient = \@quotient;
  my @answer;
  my $remainder;
  do {
    ($quotient,$remainder) = dividebybase($quotient,$base);
# these commented out print statements are for convert.t DO NOT REMOVE!
#$used{$remainder} = 1;
#print $remainder;
#print "    *" if $remainder > 86;
#print "\n";
    unshift @answer, $ary->[$remainder];
  } while grep {$_} @$quotient;

#foreach (sort {$b <=> $a} keys %used) {
#print " $_,\n" if $used{$_} && $_ > 85;
#print "\t$_\t=> \n" if !$used{$_} && $_ < 86;
#}
  join '', @answer;
}

1;

__END__

=head1 NAME

Math::Base::Convert::CalcPP

=head1 DESCRIPTION

This module contains the standard methods used by B<Math::Base::Convert> to
convert from one base number to another base number.

=over 4

=item * $carry = addbaseno($reg32ptr,$int)

This function adds an integer < 65536 to a long n*32 bit register and
returns the carry.

=item * multiply($reg32ptr,$int)

This function multiplies a long n*32 bit register by an integer < 65536

=item * ($qptr,$remainder) = dividebybase($reg32ptr,$int)

this function divides a long n*32 bit register by an integer < 65536 and
returns a pointer to a long n*32 bit quotient and an integer remainder.

=item * $bc->useFROMbaseto32wide

This method converts FROM an input base string to a long n*32 bit register using
an algorithim like:

	$longnum = 0;
	for $char ( $in_str =~ /./g ) {
	  $longnum *= $base;
	  $longnum += $value{$char)
	}
	return $number;

=item * $output = $bc->use32wideTObase

This method converts a long n*32 bit register TO a base number using an
algorithim like:

    $output = '';
    while( $longnum > 0 ) {
      $output = ( $longnum % $base ) . $output;
      $num = int( $longnum / $base );
    }
    return $output;

=head1 AUTHOR
   
Michael Robinton, michael@bizsystems.com

=head1 COPYRIGHT

Copyright 2012, Michael Robinton

This program is free software; you may redistribute it and/or modify it
under the same terms as Perl itself.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 
=cut

1;


lib/Math/Base/Convert/Bases.pm
#!/usr/bin/perl

package Math::Base::Convert::Bases;

$VERSION = 0.02;

package Math::Base::Convert;	# into the main package

@BASES = qw( bin dna DNA oct dec hex HEX b62 b64 m64 iru url rex id0 id1 xnt xid b85 ascii );

$signedBase = 16;               # largest allowable known signed base

my $package = __PACKAGE__;
my $packageLen = length __PACKAGE__;

sub _class {
  (my $class = (caller(1))[3]) =~ s/([^:]+)$/_bs::$1/;
  $class;
}

my $callname = __PACKAGE__ . '::_bs::';

# return a pointer to a sub for the array blessed into Package::sub::name
#

my $_bin = bless ['0', '1'], $callname . 'bin';
my $_dna = bless [qw( a c t g )], $callname . 'dna';
my $_DNA = bless [qw( A C T G )], $callname . 'DNA';
my $_ocT = bless ['0'..'7'], $callname . 'ocT';
my $_dec = bless ['0'..'9'], $callname . 'dec';
my $_heX = bless ['0'..'9', 'a'..'f'], $callname . 'heX';
my $_HEX = bless ['0'..'9', 'A'..'F'], $callname . 'HEX';
my $_b62 = bless ['0'..'9', 'a'..'z', 'A'..'Z'], $callname . 'b62';
my $_b64 = bless ['0'..'9', 'A'..'Z', 'a'..'z', '.', '_'], $callname . 'b64';
my $_m64 = bless ['A'..'Z', 'a'..'z', '0'..'9', '+', '/'], $callname . 'm64';
my $_iru = bless ['A'..'Z', 'a'..'z', '0'..'9', '[', ']'], $callname . 'iru';
my $_url = bless ['A'..'Z', 'a'..'z', '0'..'9', '*', '-'], $callname . 'url';
my $_rex = bless ['A'..'Z', 'a'..'z', '0'..'9', '!', '-'], $callname . 'rex';
my $_id0 = bless ['A'..'Z', 'a'..'z', '0'..'9', '_', '-'], $callname . 'id0';
my $_id1 = bless ['A'..'Z', 'a'..'z', '0'..'9', '.', '_'], $callname . 'id1';
my $_xnt = bless ['A'..'Z', 'a'..'z', '0'..'9', '.', '-'], $callname . 'xnt';
my $_xid = bless ['A'..'Z', 'a'..'z', '0'..'9', '_', ':'], $callname . 'xid';
my $_b85 = bless ['0'..'9', 'A'..'Z', 'a'..'z', '!', '#',   # RFC 1924 for IPv6 addresses, might need to return Math::BigInt objs
            '$', '%', '&', '(', ')', '*', '+', '-', ';', '<', '=', '>', '?', '@', '^', '_', '`', '{', '|', '}', '~'], $callname . 'b85';

my $_ascii = bless [
	' ','!','"','#','$','%','&',"'",'(',')','*','+',',','-','.','/',
	'0','1','2','3','4','5','6','7','8','9',
	':',';','<','=','>','?','@',
	'A','B','C','D','E','F','G','H','I','J','K','L','M',
	'N','O','P','Q','R','S','T','U','V','W','X','Y','Z',
	'[','\\',']','^','_','`',
	'a','b','c','d','e','f','g','h','i','j','k','l','m',
	'n','o','p','q','r','s','t','u','v','w','x','y','z',
	'{','|','}','~'], $callname . 'ascii';		# 7 bit printable ascii, base 96

#my $_ebcdic = bless [qw
#	( 0  1  2  3 37 2D 2E 2F 16  5 25 0B 0C 0D 0E 0F 10 11 12 13 3C 3D 32 26 18 19 3F 27 1C 1D 1E 1F 
#	 40 4F 7F 7B 5B 6C 50 7D 4D 5D 5C 4E 6B 60 4B 61 F0 F1 F2 F3 F4 F5 F6 F7 F8 F9 7A 5E 4C 7E 6E 6F 
#	 7C C1 C2 C3 C4 C5 C6 C7 C8 C9 D1 D2 D3 D4 D5 D6 D7 D8 D9 E2 E3 E4 E5 E6 E7 E8 E9 4A E0 5A 5F 6D 
#	 79 81 82 83 84 85 86 87 88 89 91 92 93 94 95 96 97 98 99 A2 A3 A4 A5 A6 A7 A8 A9 C0 6A D0 A1  7 
#	 20 21 22 23 24 15 6 17 28 29 2A 2B 2C 9 0A 1B 30 31 1A 33 34 35 36 8 38 39 3A 3B  4 14 3E E1 41 
#	 42 43 44 45 46 47 48 49 51 52 53 54 55 56 57 58 59 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 
#	 77 78 80 8A 8B 8C 8D 8E 8F 90 9A 9B 9C 9D 9E 9F A0 AA AB AC AD AE AF B0 B1 B2 B3 B4 B5 B6 B7 B8
#	 B9 BA BB BC BD BE BF CA CB CC CD CE CF DA DB DC DD DE DF EA EB EC ED EE EF FA FB FC FD FE FF)], $callname . 'ebcdic';

sub bin { $_bin }
sub dna { $_dna }
sub DNA { $_DNA }
sub ocT { $_ocT }
sub dec { $_dec }
sub heX { $_heX }
sub HEX { $_HEX }
sub b62 { $_b62 }
sub b64 { $_b64 }
sub m64 { $_m64 }
sub iru { $_iru }
sub url { $_url }
sub rex { $_rex }
sub id0 { $_id0 }
sub id1 { $_id1 }
sub xnt { $_xnt }
sub xid { $_xid }
sub b85 { $_b85 }
sub ascii { $_ascii }
#sub ebcdic { $_ebcdic }

# Since we're not using BIcalc, the last test can be eliminated...
################### special treatment for override 'hex' ##################################

sub hex {
		#		unless	our package	and is a BC ref	and not a BI number (which is an ARRAY)
  unless (ref($_[0]) && $package eq substr(ref($_[0]),0,$packageLen) && (local *glob = $_[0]) && *glob{HASH}) {
# $package, $filename, $line, $subroutine, $hasargs
#       0       1        2         3            4
#   if   defined      and    hasargs
    if ( defined $_[0] && (caller(0))[4] ) {
      return CORE::hex $_[0];
    }
  }
  return heX();
}

################### special treatment for override 'oct' #################################
sub oct {
		#		unless	our package	and is a BC ref	and not a BI number (which is an ARRAY)
  unless (ref($_[0]) && $package eq substr(ref($_[0]),0,$packageLen) && (local *glob = $_[0]) && *glob{HASH}) {
# $package, $filename, $line, $subroutine, $hasargs
    #       0       1        2         3            4
#   if   defined      and    hasargs
    if ( defined $_[0] && (caller(0))[4] ) {
      return CORE::oct $_[0];
    }
  }
  return ocT();
}

################################## REMOVE ABOVE CODE ###################

# return a hash of all base pointers
#
sub _bases {
  no strict;
  my %bases;
  foreach (@BASES) {
    my $base = $_->();
    ref($base) =~ /([^:]+)$/;
    $bases{$1} = $base;
  }
  \%bases;
}
1;

__END__

=head1 NAME - Math::Base::Convert::Bases

This package contains no documentation

See L<Math::Base::Convert> instead

=head1 AUTHOR

Michael Robinton, michael@bizsystems.com

=head1 COPYRIGHT

Copyright 2012, Michael Robinton

This program is free software; you may redistribute it and/or modify it
under the same terms as Perl itself.   

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

=cut 

1;
